scale-out-relay/src/main/scala/com/normation/plugins/scaleoutrelay/ScaleOutRelayService.scala

```diff
@@ -0,0 +1,192 @@
+package com.normation.plugins.scaleoutrelay
+
+import bootstrap.liftweb.RudderConfig
+import com.normation.cfclerk.domain.TechniqueVersion
+import com.normation.errors.IOResult
+import com.normation.errors._
+import com.normation.eventlog.EventActor
+import com.normation.eventlog.ModificationId
+import com.normation.inventory.domain.NodeId
+import com.normation.rudder.domain.nodes.Node
+import com.normation.rudder.domain.nodes.NodeGroup
+import com.normation.rudder.domain.nodes.NodeGroupCategoryId
+import com.normation.rudder.domain.nodes.NodeGroupId
+import com.normation.rudder.domain.nodes.NodeInfo
+import com.normation.rudder.domain.policies._
+import com.normation.rudder.domain.queries._
+import com.normation.rudder.rule.category.RuleCategoryId
+import net.liftweb.common.EmptyBox
+import net.liftweb.common.Failure
+import net.liftweb.common.Full
+import zio.ZIO
+
+
+object ScaleOutRelayService {
+  val SYSTEM_GROUPS = "SystemGroups"
+  val DISTRIBUTE_POLICY = "distributePolicy"
+  val COMMON = "common"
+
+  private val nodeInfosService = RudderConfig.nodeInfoService
+  private val woLDAPNodeGroupRepository = RudderConfig.woNodeGroupRepository
+  private val woLDAPNodeRepository = RudderConfig.woNodeRepository
+  private val woDirectiveRepository = RudderConfig.woDirectiveRepository
+  private val woRuleRepository = RudderConfig.woRuleRepository
+  private val uuidGen = RudderConfig.stringUuidGenerator
+  private val policyServerManagementService = RudderConfig.policyServerManagementService
+  private val removeNodeService = RudderConfig.removeNodeService
+
+  def promoteNodeToRelay(uuid: NodeId, actor: EventActor, reason:Option[String]): ZIO[Any, RudderError, NodeInfo] = {
+    val modId = ModificationId(uuidGen.newUuid)
+    for {
+      nodeInfos   <- getNodeToPromote(uuid)
+      updatedNode =  createNodeToPolicyServerType(nodeInfos)
+      _           <- removeNodeService.removeNode(nodeInfos.node.id, modId, actor).toIO
+      newRelay <- createRelayFromNode(updatedNode, modId, actor, reason)
```

Comment:
I think that if that fails, we should try to restaure old node as it was. IE we should keep original node entry around, and in case of error in create, try to delete what was already saved and save back original entry. What do you thing @VinceMacBuche ?

Reply:
I think i agree with you  @fanf

@ElaadF You can use ZIO.catchall on each line of the for and if it fails, call a  function that remove the current element, (or the previous one), which call a function that remove element added before until you restore the node to it's previous state

Ie, if the order of action is 

transform node to relay
add Group
add Directive
add Rule

and if an error occurs while you are adding the directive, it should

remove the directive
remove the group
transform the relay to a normal node

(looks like demote function though)

Reply:
this things should already be in node deletion. That can/should be factored together (up to node removal)